compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala

```diff
@@ -127,7 +127,20 @@ trait ConstraintHandling {
     * of some param when comparing types might lead to infinite recursion. Consider `bounds` instead.
     */
   def fullBounds(param: TypeParamRef)(using Context): TypeBounds =
-    nonParamBounds(param).derivedTypeBounds(fullLowerBound(param), fullUpperBound(param))
```

Comment:
This is likely still not the correct place to fix this, constraints shouldn't be cyclic (although the assert for this is disabled by default for performance: https://github.com/lampepfl/dotty/blob/3c8200671995ea75b97b493081cc4e51109cfd04/compiler/src/dotty/tools/dotc/config/Config.scala#L21-L24), it's possible the logic in GadtConstraint.scala violates this.

Reply:
Thanks, I'll check.

Reply:
Enabling doesn't seem to trigger any errors.  Are we sure it's a cycle in the constraints?  Need to investigate more.

Reply:
So inference first creates the type refs while typing the patterns (in `maximizeType` via `typedUnApply`) as `F$1 <: F` and `F$2 <: F$1`.  Just a little later runs over the patterns again in `indexPattern` where it calls `fullBounds` leading to `F$1(param)1 >: F$2 <: F` and `F$2(param)1 <: F$1`.

So is it circular constraints? 🤔

Reply:
Yeah, either gadt.fullBounds should not produce cycles like this, or we shouldn't use it here, wdyt @abgruszecki ?

Reply:
The way I see it, `fullLowerBound` aren't wrong in their answers.  But put together they lead to a mutual reference.  So fixing this in fullBounds by inspecting the results seems reasonable to me.

Hopefully I've fixed the test suite now.

Reply:
I don't know. 🤷🏼‍♂️   But GadtConstraint redefines its fullLowerBound and fullUpperBound so possibly not?